### PR TITLE
Add informer for statefulset workloads

### PR DESCRIPTION
**Pull request by savitaashture**
# Changes

Added informers to watch changes for statefulesets workload

Signed-off-by: Savita Ashture <sashture@redhat.com>

**Verification:**

1. Deleting pod related to statefulset will be recreated by statefulset Object as statefulset is a owner for pod

Screenshot from 2023-08-30 14-52-34

2. Deleting statefulset object _**now recreated**_ because of the _**statefulsetInformer changes** **earlier it was not happening**_

Screenshot from 2023-08-30 15-12-27

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Run `make test lint` before submitting a PR
- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```release-note
statefulSets watches any changes to statefulSet workload
```

### CODE REVIEW

**Comment by savitaashture** [comment]
Will reverify this PR changes once after this issue fix https://github.com/tektoncd/operator/issues/1663

**Review by khrm — COMMENTED** [review]
/approve

Needed for results DB or any DB we use in the future. DBs should be stateful-set not deployment.

**Comment by savitaashture** [comment]
> Need to handle changes when statefulsets updated directly

Done

**Comment by savitaashture** [comment]
**Pending:**
1. E2E need to be added
2. Code Refactor to see if can be avoided duplication
But will be done in follow ups PR


PR is ready for review cc @tektoncd/operator-maintainers

File: pkg/reconciler/kubernetes/tektoninstallerset/install.go [hunk 1]
```diff
@@ -262,6 +351,91 @@ func (i *installer) updateDeployment(existing *unstructured.Unstructured, existi
...
 }
 
+func (i *installer) ensureStatefulSet(expected *unstructured.Unstructured) error {
+	i.logger.Debugw("verifying a statefulset",
+		"name", expected.GetName(),
+		"namespace", expected.GetNamespace(),
+	)
+
+	// update proxy settings
+	err := common.ApplyProxySettings(expected)
```
Comment (jkandasa) [inline pkg/reconciler/kubernetes/tektoninstallerset/install.go:361]: inside `ApplyProxySettings` we have a check. that will ignore `StatefulSet`,
```
if u.GetKind() != "Deployment" {
```

Do we really need proxy things to StatefulSets?

Comment (savitaashture) [reply pkg/reconciler/kubernetes/tektoninstallerset/install.go:361]: I am not sure do we really need it but added considering what if require by someone in disconnected cluster and all

But ya i need to add condition for statefulset

**Comment by piyush-garg** [comment]
why we need a specific condition for stateful set? just adding informers should also work ?

**Comment by savitaashture** [comment]
> why we need a specific condition for stateful set? just adding informers should also work ?

adding informer will work if user manually delete statefulset

but what if user edit statefulset manually then tektoninstallerset should revert those changes similar to Deployment 
thats why added other condition along with informers

**Comment by jkandasa** [comment]
@piyush-garg for now statefulset allows the user to control replicas count. So we need a special logic to handle that.
May be in the future, we can revert the replicas count from the user. 

and we have an open issue https://github.com/tektoncd/operator/issues/1672, when we fix this and that time if replicas managed by operator, we can remove these changes(this PR)

**Comment by piyush-garg** [comment]
lgtm
cc @jkandasa

**Comment by jkandasa** [comment]
Thanks @savitaashture !

**Pull request merged**